fastfield_codecs/src/ip_codec.rs

```diff
@@ -0,0 +1,760 @@
+/// This codec takes a large number space (u128) and reduces it to a compact number space.
+///
+/// It will find spaces in the numer range. For example:
+///
+/// 100, 101, 102, 103, 104, 50000, 50001
+/// could be mapped to
+/// 100..104 -> 0..4
+/// 50000..50001 -> 5..6
+///
+/// Compact space 0..6 requires much less bits than 100..50001
+///
+/// The codec is created to compress ip addresses, but may be employed in other use cases.
+use std::{
+    cmp::Ordering,
+    collections::BinaryHeap,
+    io::{self, Write},
+    net::{IpAddr, Ipv6Addr},
+    ops::RangeInclusive,
+};
+
+use tantivy_bitpacker::{self, BitPacker, BitUnpacker};
+
+use crate::FastFieldCodecReaderU128;
+
+pub fn ip_to_u128(ip_addr: IpAddr) -> u128 {
+    let ip_addr_v6: Ipv6Addr = match ip_addr {
+        IpAddr::V4(v4) => v4.to_ipv6_mapped(),
+        IpAddr::V6(v6) => v6,
+    };
+    u128::from_be_bytes(ip_addr_v6.octets())
+}
+
+const INTERVALL_COST_IN_BITS: usize = 64;
+
+#[derive(Default, Debug)]
+pub struct IntervalEncoding();
+
+pub struct IntervalCompressor {
+    pub null_value: u128,
+    min_value: u128,
+    max_value: u128,
+    compact_space: CompactSpace,
+    pub num_bits: u8,
+}
+
+const STOP_BIT: u8 = 128u8;
+
+fn serialize_vint(mut val: u128, output: &mut Vec<u8>) {
```

Comment:
Can we move this function to the vint stuff in tantivy_common? I think there is already a deps fastfield_codecs -> tantivy-common anyway